library/items/item-group/src/main/java/org/quiltmc/qsl/itemgroup/mixin/ItemGroupMixin.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.itemgroup.mixin;
+
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Mutable;
+import org.spongepowered.asm.mixin.Shadow;
+
+import net.minecraft.item.ItemGroup;
+
+import org.quiltmc.qsl.itemgroup.impl.ItemGroupExtensions;
+
+@Mixin(ItemGroup.class)
+public abstract class ItemGroupMixin implements ItemGroupExtensions {
+	@Shadow
+	@Final
+	@Mutable
+	public static ItemGroup[] GROUPS;
+
+	@Override
+	public void expandArray() {
```

Comment:
I think that we should have a registry for item groups, instead of growing the item group array each time an item group is created. That way we can have all of the item groups in one place, and register them all at once in one go.
Also I feel that registring an item group *implicitly* when it's created is a bit unintuitive and non-conforming when most of the rest of the API requires explicit registration.

Reply:
I believe the idea with modifying the array each time is so that mods can inspect all the tabs as they would in vanilla, or even add tabs themselves without the standard library.

Reply:
>so that mods can inspect all the tabs as they would in vanilla
My idea is to have the QSL expand and populate the array in bulk and on demand. Maybe we could update the vanilla "mirror" every time it's accessed and the QSL registry changes, and that way it syncs up.

>or even add tabs themselves without the standard library.
Sure, modders can just use mixins to expand the array this way, but the standard library need not to do that - I would expect that most modders would use the QSL to solve problems that have already been solved before (i.e. not reinventing the wheel), so IMO should a mod reimplement something that already exists in the QSL, it's entirely on its own to be compatible.

tl;dr a mod can do whatever it likes, but it's not the QSL's problem if they are incompatible

Reply:
A registry implies removals are also valid, so any implementation would need to take that into account.

Reply:
A full-on registry is likely unnecessary. Lazy initialization of item groups should be sufficient.

Reply:
Yeah, dynamic removal of item groups is likely unnessicary